service/service.go

```diff
@@ -0,0 +1,181 @@
+package service
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+
+	"github.com/ONSdigital/dp-dimension-extractor/dimension"
+	"github.com/ONSdigital/dp-dimension-extractor/schema"
+	"github.com/ONSdigital/go-ns/avro"
+	"github.com/ONSdigital/go-ns/kafka"
+	"github.com/ONSdigital/go-ns/log"
+	"github.com/ONSdigital/go-ns/s3"
+)
+
+// ----------------------------------------------------------------------------
+
+// Service represents the necessary config for dp-dimension-extractor
+type Service struct {
+	EnvMax       int64
+	Consumer     *kafka.ConsumerGroup
+	ImportAPIURL string
+	Producer     kafka.Producer
+	S3           *s3.S3
+}
+
+type inputFileAvailable struct {
+	FileURL    string `avro:"file_url"`
+	InstanceID string `avro:"instance_id"`
+}
+
+type dimensionExtracted struct {
+	FileURL    string `avro:"file_url"`
+	InstanceID string `avro:"instance_id"`
+}
+
+// ----------------------------------------------------------------------------
+
+// Start handles consumption of events ad produces new events
+func (svc *Service) Start() {
+	log.Info("application started", log.Data{"import-api-url": svc.ImportAPIURL})
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, os.Kill)
+
+	healthChannel := make(chan bool)
+	exitChannel := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-c:
+				//Falls into this block when the service is shutdown to safely close the consumer
+				log.Info("safely closing consumer and producer on service shutdown", nil)
+				svc.Consumer.Closer <- true
+				svc.Producer.Closer <- true
+				exitChannel <- true
+				return
+			case message := <-svc.Consumer.Incoming:
+				m, err := readMessage(message.GetData())
+				if err != nil {
+					log.ErrorC("could not unmarshal message", err, log.Data{"instanceID": m.InstanceID})
+					continue
+				}
+				log.Info("received message", log.Data{"file-url": m.FileURL, "instance-id": m.InstanceID})
+
+				// Get csv from S3 bucket using m.S3URL
+				//s3Value := svc.s3Value
+				file, err := svc.S3.Get(m.FileURL)
+				if err != nil {
+					log.ErrorC("encountered error retrieving csv file", err, log.Data{"instanceID": m.InstanceID})
+					continue
+				}
+
+				log.Info("file successfully read from aws", log.Data{"instanceID": m.InstanceID})
+
+				csvReader := csv.NewReader(file)
```

Comment:
This case is quite long - from here might be a good opportunity to break out some methods. The schemas can be created further down where they're actually needed, in a method call. CSV handling could be its own method(s)